File: src/Floss.App/Brushes/CurveOption.cs
using System;
using System.Collections.Generic;

namespace Floss.App.Brushes;

public enum SensorCombineMode { Multiply, Add }

// Maps one or more sensor inputs through cubic curves to produce a [0,1] multiplier.
// Disabled or empty → returns 1.0 (no effect on the target parameter).
public sealed class CurveOption
{
    public bool             IsEnabled   { get; set; } = true;
    public float            Strength    { get; set; } = 1.0f;
    public float            MinOutput   { get; set; } = 0.0f;
    public float            MaxOutput   { get; set; } = 1.0f;
    public SensorCombineMode CombineMode { get; set; } = SensorCombineMode.Multiply;
    public List<SensorConfig> Sensors   { get; set; } = [];

    // Returns a multiplier in approximately [MinOutput, MaxOutput], lerped toward 1.0
    // by (1 - Strength). No sensors / disabled → 1.0.
    public float Compute(in StrokePoint sp)
    {
        if (!IsEnabled || Sensors.Count == 0) return 1.0f;

        float combined;
        if (CombineMode == SensorCombineMode.Multiply)
        {
            combined = 1.0f;
            foreach (var s in Sensors) combined *= s.CurvedValue(sp);
        }
        else
        {
            combined = 0.0f;
            foreach (var s in Sensors) combined += s.CurvedValue(sp);
            combined = Math.Clamp(combined / Sensors.Count, 0, 1);
        }

        float raw = MinOutput + (MaxOutput - MinOutput) * combined;
        return 1.0f - Strength + Strength * raw;
    }

    public CurveOption Clone()
    {
        var c = new CurveOption
        {
            IsEnabled = IsEnabled, Strength = Strength,
            MinOutput = MinOutput, MaxOutput = MaxOutput,
            CombineMode = CombineMode
        };
        foreach (var s in Sensors) c.Sensors.Add(s.Clone());
        return c;
    }

    // ── Factories ────────────────────────────────────────────────────────────

    public static CurveOption Off() => new() { IsEnabled = false };

    public static CurveOption Pressure(float gamma, float min = 0f, float max = 1f)
    {
        var opt = new CurveOption { MinOutput = min, MaxOutput = max };
        opt.Sensors.Add(new SensorConfig { Type = SensorType.Pressure, Curve = GammaCurve(gamma) });
        return opt;
    }

    // Pressure + inverse-speed combined (size shrinks at high velocity).
    public static CurveOption PressureSpeed(float gamma, float velStrength, float min = 0f, float max = 1f)
    {
        var opt = new CurveOption { MinOutput = min, MaxOutput = max };
        opt.Sensors.Add(new SensorConfig { Type = SensorType.Pressure, Curve = GammaCurve(gamma) });
        opt.Sensors.Add(new SensorConfig { Type = SensorType.Speed,    Curve = InverseSpeedCurve(velStrength) });
        return opt;
    }

    private static CubicCurve GammaCurve(float gamma)
    {
        const int steps = 9;
        var pts = new CurvePoint[steps];
        for (int i = 0; i < steps; i++)
        {
            float x = i / (float)(steps - 1);
            pts[i] = new CurvePoint(x, Math.Clamp(MathF.Pow(x, gamma), 0, 1));
        }
        var c = new CubicCurve();
        c.SetPoints(pts);
        return c;
    }

    // x=0 (slow) → 1.0, x=1 (fast) → (1-strength)
    private static CubicCurve InverseSpeedCurve(float strength)
    {
        var c = new CubicCurve();
        c.SetPoints([new(0f, 1f), new(1f, Math.Clamp(1f - strength, 0.05f, 1f))]);
        return c;
    }
}

File: crates/floss-brush/src/curve_option.rs
//! CurveOption — combines sensors into a single multiplier via multiply or add.



use crate::curve::CubicCurve;
use crate::sensor::{SensorConfig, SensorType};
use crate::stroke::StrokePoint;

/// How multiple sensors are combined.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum SensorCombineMode {
    /// Multiply all sensor values together.
    Multiply,
    /// Average all sensor values.
    Add,
}

/// Maps one or more sensors through cubic curves to produce a [min_output, max_output]
/// multiplier. Disabled or empty → returns 1.0 (no effect).
#[derive(Clone, Debug)]
pub struct CurveOption {
    pub enabled: bool,
    /// Lerp factor: 0 = no effect, 1 = full curve effect.
    pub strength: f32,
    /// Minimum output multiplier.
    pub min_output: f32,
    /// Maximum output multiplier.
    pub max_output: f32,
    /// How to combine multiple sensors.
    pub combine_mode: SensorCombineMode,
    /// The sensors driving this option.
    pub sensors: Vec<SensorConfig>,
}

impl CurveOption {
    /// Create a disabled option (returns 1.0).
    pub fn off() -> Self {
        Self {
            enabled: false,
            strength: 1.0,
            min_output: 0.0,
            max_output: 1.0,
            combine_mode: SensorCombineMode::Multiply,
            sensors: Vec::new(),
        }
    }

    /// Create a simple pressure-based option with a power curve.
    pub fn pressure(gamma: f32, min: f32, max: f32) -> Self {
        let mut opt = Self {
            enabled: true,
            strength: 1.0,
            min_output: min,
            max_output: max,
            combine_mode: SensorCombineMode::Multiply,
            sensors: Vec::new(),
        };
        opt.sensors.push(SensorConfig {
            sensor_type: SensorType::Pressure,
            curve: gamma_curve(gamma),
            length: 1000.0,
        });
        opt
    }

    /// Evaluate the combined sensor value.
    ///
    /// Returns a multiplier in approximately [min_output, max_output],
    /// lerped toward 1.0 by (1 - strength). No sensors or disabled → 1.0.
    pub fn compute(&self, sp: &StrokePoint) -> f32 {
        if !self.enabled || self.sensors.is_empty() {
            return 1.0;
        }

        let combined = match self.combine_mode {
            SensorCombineMode::Multiply => {
                let mut v = 1.0f32;
                for s in &self.sensors {
                    v *= s.curved_value(sp);
                }
                v
            }
            SensorCombineMode::Add => {
                let sum: f32 = self.sensors.iter().map(|s| s.curved_value(sp)).sum();
                (sum / self.sensors.len() as f32).clamp(0.0, 1.0)
            }
        };

        let raw = self.min_output + (self.max_output - self.min_output) * combined;
        1.0 - self.strength + self.strength * raw
    }
}

/// Build a power-curve (x^gamma) as a CubicCurve with 9 sample points.
fn gamma_curve(gamma: f32) -> CubicCurve {
    const STEPS: usize = 9;
    let mut c = CubicCurve::identity();
    let pts: Vec<_> = (0..STEPS)
        .map(|i| {
            let x = i as f32 / (STEPS - 1) as f32;
            crate::curve::CurvePoint::new(x, x.powf(gamma).clamp(0.0, 1.0))
        })
        .collect();
    c.set_points(pts);
    c
}
